flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/partitioning/DistributedPrePartitionOperator.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.flink.cdc.common.event.SchemaChangeEvent;
+import org.apache.flink.cdc.common.event.TableId;
+import org.apache.flink.cdc.common.function.HashFunction;
+import org.apache.flink.cdc.common.function.HashFunctionProvider;
+import org.apache.flink.cdc.common.schema.Schema;
+import org.apache.flink.cdc.common.utils.SchemaUtils;
+import org.apache.flink.cdc.runtime.operators.schema.regular.SchemaOperator;
+import org.apache.flink.cdc.runtime.serializer.event.EventSerializer;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.ChainingStrategy;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Operator for processing events from {@link SchemaOperator} before {@link EventPartitioner} with
+ * distributed topology.
+ */
+@Internal
+public class DistributedPrePartitionOperator extends AbstractStreamOperator<PartitioningEvent>
+        implements OneInputStreamOperator<Event, PartitioningEvent>, Serializable {
+    private static final long serialVersionUID = 1L;
+
+    private final int downstreamParallelism;
+    private final HashFunctionProvider<DataChangeEvent> hashFunctionProvider;
+
+    // Schema and HashFunctionMap used in schema inferencing mode.
+    private transient Map<TableId, Schema> schemaMap;
+    private transient Map<TableId, HashFunction<DataChangeEvent>> hashFunctionMap;
+
+    private transient int subTaskId;
+
+    public DistributedPrePartitionOperator(
+            int downstreamParallelism, HashFunctionProvider<DataChangeEvent> hashFunctionProvider) {
+        this.chainingStrategy = ChainingStrategy.ALWAYS;
+        this.downstreamParallelism = downstreamParallelism;
+        this.hashFunctionProvider = hashFunctionProvider;
+    }
+
+    @Override
+    public void open() throws Exception {
+        super.open();
+        subTaskId = getRuntimeContext().getTaskInfo().getIndexOfThisSubtask();
+        schemaMap = new HashMap<>();
+        hashFunctionMap = new HashMap<>();
+    }
+
+    @Override
+    public void processElement(StreamRecord<Event> element) throws Exception {
+        Event event = element.getValue();
+        if (event instanceof SchemaChangeEvent) {
+            SchemaChangeEvent schemaChangeEvent = (SchemaChangeEvent) event;
+            TableId tableId = schemaChangeEvent.tableId();
+
+            // Update schema map
+            schemaMap.compute(
+                    tableId,
+                    (tId, oldSchema) ->
+                            SchemaUtils.applySchemaChangeEvent(oldSchema, schemaChangeEvent));
+
+            // For malformed dangling dropTableEvents, we simply ignore this event to avoid breaking
+            // the pipeline.
```

Comment:
I remember it came from an ancient failure case when we're trying to start from a previous Binlog position:

```
Binlog ----> Create Table A --------> Drop Table A -----------> Now ------>
                                ^
                                |
                      Starting up from here
```

Since we're fetching latest table schema when job starts ("Now"), Table A will not be known since it has been dropped before. However, by reading binlog after CreateTableEvent and before DropTableEvent in the binlog stream, we will come up with a vexing situation, as it is trying to drop an unknown table.

Reply:
Sorry, they're some residual coming from previous attempt to create a unified topology for both regular and distributed sources. Since this operator will only be used in the new topology, the dangling "DropTableEvent" case is impossible.